### PR TITLE
WT-6911 Write "block manager" subpage for Architecture Guide

### PR DESCRIPTION
**Pull request by tetsuo-cpp**

### CODE REVIEW

File: src/docs/arch-block.dox [hunk 1]
```diff
@@ -1,9 +1,132 @@
...
+Internally, WiredTiger uses a data structure called an extent list or a
+\c WT_EXTLIST to track file usage. An extent list consists of a series of
+extents (or \c WT_EXT elements). Each extent tracks a file offset and size to
+track a portion of the file.
+
+There are three extent lists that are maintained per checkpoint:
+
+- \c alloc: The extents allocated for a given checkpoint.
+- \c avail: The extents that are unused and available for allocation.
+- \c discard: The extents freed in the current checkpoint. After the checkpoint
```
Comment (tetsuo-cpp) [inline src/docs/arch-block.dox:81]: @keithbostic I'm a bit hazy about this part. Is this how it works?

This is what makes sense to me but I wasn't able to see an obvious part of code where we do that.

File: src/docs/arch-block.dox [hunk 2]
```diff
@@ -1,9 +1,132 @@
...
+
+For example, if we specify an \c allocation_size of \c 4KB, blocks of size \c 8KB
+and \c 12KB would be permitted but NOT \c 10KB. Higher allocation sizes may
+yield better compression rates, yet may cause larger page sizes. The allocation
+size is set to \c 4KB by default which is a good choice unless your OS or storage
+device has special requirements.
+
+@subsection block_compression Block compression
+
+The \c block_compressor configuration controls the compression algorithm used
```
Comment (tetsuo-cpp) [inline src/docs/arch-block.dox:110]: Was considering adding a snippet showing the `WT_COMPRESSOR` struct and the methods required (and the same for `WT_ENCRYPTOR` below). Do we think that's useful?

File: src/docs/arch-block.dox (hunk 1 above)
Comment (keithbostic) [reply src/docs/arch-block.dox:81]: Yes, that's correct as best I recall. Those are the major groupings, certainly.

File: src/docs/arch-block.dox (hunk 2 above)
Comment (keithbostic) [reply src/docs/arch-block.dox:110]: Encryption & compression are almost entirely handled in the upper, Btree layer, not the block manager.

File: src/docs/arch-block.dox [hunk 3]
```diff
@@ -1,9 +1,132 @@
...
-are configured.
+The WiredTiger block manager subsystem manages the reading and writing of data
+from the disk. It is designed to facilitate high performance, economic use of
+disk space and customisability.
+
+@section block What is a block?
+
+The block manager API introduces the concept of a "block" which represents a
+chunk of data that is stored on the disk and operated on as a single atomic
+unit. Each WiredTiger data file (any file in the home directory with the \c .wt
```
Comment (keithbostic) [inline src/docs/arch-block.dox:11]: "atomic" bothers me a little, I'm not sure why you're using that word, and it has other meanings around data structure that don't apply here?

File: src/docs/arch-block.dox [hunk 4]
```diff
@@ -1,9 +1,132 @@
...
+suffix) is made up of these blocks.
+
+Since a data file can contain page images from multiple reconciliations of the
+same page, each block is associated with a single checkpoint. The API is
```
Comment (keithbostic) [inline src/docs/arch-block.dox:15]: A block can appear in any number of checkpoints, that is, writing a new checkpoint doesn't rewrite all of the blocks it contains.

File: src/docs/arch-block.dox [hunk 5]
```diff
@@ -1,9 +1,132 @@
...
+designed in a way that allows users to load a checkpoint into the block manager
+and then read the blocks from that checkpoint as if they were a contiguous
+stream of chunks in the file (in reality, blocks from different checkpoints can
+be interleaved).
```
Comment (keithbostic) [inline src/docs/arch-block.dox:19]: There's no API that reads blocks as a contiguous stream? I'm not sure what you're trying to say here.

File: src/docs/arch-block.dox [hunk 6]
```diff
@@ -1,9 +1,132 @@
...
+
+@section block_usage How WiredTiger uses the block manager
+
+Whenever a page is read into the cache, WiredTiger will read its corresponding
+block from the data file. This block contains what's referred to as the
+"page image" in other subsystems and has the newest versions of data for each
+key.
+
+Conversely, when a page is reconciled whether that is due to a page eviction or
```
Comment (keithbostic) [inline src/docs/arch-block.dox:28]: Depending on the target of this document, you might want to define "reconciliation", it's not a general industry term as far as I know.

File: src/docs/arch-block.dox [hunk 7]
```diff
@@ -1,9 +1,132 @@
...
+by a checkpoint, a new page image is constructed and it is written as a new
+block in the data file. In the case of checkpoint, the newly written block is
```
Comment (keithbostic) [inline src/docs/arch-block.dox:30]: Here's where we might want to mention the "no overwrite" nature of the block manager?

File: src/docs/arch-block.dox [hunk 8]
```diff
@@ -1,9 +1,132 @@
...
+bound to the current checkpoint and in the case of eviction, it is associated
+with the next checkpoint.
```
Comment (keithbostic) [inline src/docs/arch-block.dox:32]: This isn't quite right: we write blocks for whatever reason, checkpoint or eviction. At some point, we do a checkpoint, and all blocks written up to that time appear as part of the newly written checkpoint.

File: src/docs/arch-block.dox [hunk 9]
```diff
@@ -1,9 +1,132 @@
...
+
+@section block_implementation How the block manager is implemented
+
+We've already discussed the concept of a block and how the data file's on-disk
+representation is divided into smaller blocks that are each associated with a
+checkpoint.
```
Comment (keithbostic) [inline src/docs/arch-block.dox:38]: To repeat a previous comment, blocks can be in any number of checkpoints. To summarize, imagine we write blocks 1-5, then do a checkpoint, write new blocks 6-8, then rewrite blocks 1 as block 9 and block 2 as block 10, then do a checkpoint.

The first checkpoint includes blocks 1-5, the second will include blocks 3-10. In other words, until the first checkpoint goes away, blocks 1-5 remain and appear in that first checkpoint, but the second checkpoint, because it rewrote 2 of the blocks and added a couple more blocks, will "include" a different, overlapping set of blocks.

For the purposes of this sentence, the key is that block 3 will appear in both checkpoints (and, in fact, in every checkpoint going forward until the block is rewritten or deleted).

File: src/docs/arch-block.dox [hunk 10]
```diff
@@ -1,9 +1,132 @@
...
+
+In this section, we'll dive into the file layout in more detail and talk about
+some of the core data structures used as part of block allocation.
+
+@subsection block_layout Block layout
+
+Each block consists of a page header, a block manager header followed by the
+actual page image.
```
Comment (keithbostic) [inline src/docs/arch-block.dox:46]: I'd say "a page header, followed by a block manager header, followed by ...", so it's clear there are 3 distinct pieces.

File: src/docs/arch-block.dox [hunk 11]
```diff
@@ -1,9 +1,132 @@
...
+\c WT_EXTLIST to track file usage. An extent list consists of a series of
+extents (or \c WT_EXT elements). Each extent tracks a file offset and size to
+track a portion of the file.
+
+There are three extent lists that are maintained per checkpoint:
+
+- \c alloc: The extents allocated for a given checkpoint.
+- \c avail: The extents that are unused and available for allocation.
+- \c discard: The extents freed in the current checkpoint. After the checkpoint
+completes, these extents are shifted to the \c avail extent list.
```
Comment (keithbostic) [inline src/docs/arch-block.dox:82]: Because blocks can appear in multiple checkpoints, you have to expect an allocation in one checkpoint, then lots of intermediate checkpoints using it, and then  an eventual discard in another checkpoint.

Imagine we allocate a block in checkpoint 37, do nothing at all with that block for checkpoints 38-41, then delete the block in checkpoint 42. The block will appear in checkpoint 37's alloc list, won't appear in any list in checkpoints 38-41, and then appears in checkpoint 42's discard list.

As each checkpoint is deleted, the blocks it allocated move into the next checkpoint's list (so, when I delete checkpoint 37, the blocks in its alloc list are merged into the next checkpoint's alloc list).

When a block appears on both the alloc & discard lists in a single checkpoint, it moves to the avail list, because that means no checkpoint is still using it.

In other words, once a block is allocated, it cannot be reused until it's deleted, and no intermediate checkpoints are still using it.

File: src/docs/arch-block.dox (hunk 2 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox:110]: Hmm, maybe it should be a separate page then? Encryption and compression were mentioned in the ticket.

Comment (keithbostic) [reply src/docs/arch-block.dox:110]: I don't know anything about the ticket's goals or history,  but I can't think of any code in the block manager that deals with encryption or compression. There's a minor hook that adjusts compression based on the compression state, but that's about it.

File: src/docs/arch-block.dox (hunk 3 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox:11]: I'm using the word atomic because that means that it's irreducible which is what I'm trying to convey. I can remove it as "operated on as a single unit" is probably descriptive enough by itself.

File: src/docs/arch-block.dox (hunk 10 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox]: 3536114

File: src/docs/arch-block.dox (hunk 6 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox:28]: The architecture guide is written for developers and TSEs. In any case, the team is also writing docs on checkpoint and eviction so I don't think we should cover reconciliation here.

File: src/docs/arch-block.dox (hunk 2 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox:110]: > I don't know anything about the ticket's goals or history

The goal of the project is to write documentation about the system for developers and other people who might be interested such as TSEs. The goal of this ticket is to cover the block manager portion of the system.

> but I can't think of any code in the block manager that deals with encryption or compression

We're also talking about what data gets passed to the block manager and how WiredTiger uses the API, so I'd say it still fits with this page.

File: src/docs/arch-block.dox (hunk 5 above)
Comment (tetsuo-cpp) [reply src/docs/arch-block.dox:19]: I misunderstood the `checkpoint_load` API. I'll take a closer look and revise this part.

**Review by keitharnoldsmith — COMMENTED** [review]
The other Keith has already provide a bunch of useful feedback.  So I'm going to make a few high-level points and then see where the revisions land before I start trying to suggest smaller changes.

Here are a few topics that it would be useful to provide some description of.  

- _The block manager decides where to write data_. When you write data to the block manager, the block manager chooses where to put the data and gives back an "address" indicating where it wrote the data.  This is important to discuss it is much more common to have the writer decide where the data goes—that's how files, disk drives, etc. work.
- _The address that the block manager gives back is an opaque bundle of bytes_. But it contains everything the block manager needs to find the data again when the caller asks to read that address (i.e., offset, length, checksum (I think)).  So the block manager needs to decide where to put new data, but it doesn't need to remember where it put it.  It assumes that its caller will save the address someplace and when block manager sees the address in a read, it can unpack the address to find the data.
- _The block manager is no-overwrite_. If you think about it, this is the reason why the block manager has to give back addresses for newly written data, rather than having the higher level code dictate where to put things.  If we read a block, modify it, then write (reconcile) it, we can't overwrite the original version of the block, so we have to put it someplace else. The block manager is where we know what blocks are free.  
- _How checkpoints work in terms of allocation_.  Keith B's comments already touch on this.  We can't delete an old checkpoint until a new checkpoint has been written.  We accomplish this by writing the checkpoint changes to free space. So if we crash, the old checkpoint is still there.  This is why we need be no overwrite. To be space efficient, a checkpoint only writes the blocks that have changed since the last checkpoint and then just points to the unchanged blocks in the previous checkpoint.

As I've tried to convey in my comments, I think of all these pieces as inter-related.  They aren't four different design points, they're different aspects of one design.  (At least, I assume that's how we got here; I wasn't here for the design...)

**Comment by tetsuo-cpp** [comment]
Thanks for your initial comments @keitharnoldsmith @keithbostic.

This has been deprioritized and is on the backlog so I'll close this PR. I'll reopen when it's time to work on it again.
